### DISCUSSION TITLE
Can I programmatically add to pybind11::dict ?

### DISCUSSION THREAD

**Discussion by nyue** [pybind/pybind11#5197] [Q&A] [answered]
Instead of initializing a pybind11::dict like this

```cpp
auto locals = pybind11::dict("name"_a="World", "number"_a=42);
```

Can I create an empty dictionary and add elements programmatically ?

What is the method name for inserting new item ?
```cpp
auto locals = pybind11::dict();
```

**Comment by jiwaszki** [comment, marked as answer]
Hi @nyue , it is for sure possible! You just need everything to be converted to Python objects, example:
```cpp
    m.def("get_custom_dict", []() {
        py::dict my_dict;

        my_dict[py::str("triple")] = py::str("H");
        my_dict[py::int_(3)] = py::int_(16);
        my_dict[py::int_(42)] = py::str("abc");

        return my_dict;
    });
```
Here are results in Python:
```python
In [1]: import mymodule
   ...: print(mymodule.get_custom_dict())
   ...: 
{'triple': 'H', 3: 16, 42: 'abc'}
```

Just remember that `py::dict` itself is collection of Python objects that "works on Python side", i.e. you need to capture GIL before using it. Thus all the keys and values are required to be "on Python side" as well. You can work with anything that pybind provides or you can wrap around with either `py::object` or `py::handle`. However, all basics should be covered in docs.

PS I hope that you managed to find it sooner! Cheers!

**Reply by marcingretl** [reply]
Ok,
but what if one allocates py::dict() object dynamically (via py::dict d = new py::dict()). How can I add an item in such case?

Marcin
